avoid creation of ListBuffers for Seq.empty and Seq()

mkeskells opened the pull request.

mkeskells committed:
avoid creation of ListBuffers for Seq.empty and Seq()

dwijnand commented on build.sbt:
```diff
@@ -278,6 +278,9 @@ val mimaFilterSettings = Seq {
     ProblemFilters.exclude[MissingTypesProblem]("scala.collection.immutable.Map$EmptyMap$"),
 
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Seq.empty"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Seq.apply"),
```
This breaks the forwards-compatibility guarantee.  This is the output (after removing this excludes and using MiMa latest 0.7.0):

```
[error] scala-library: Failed binary compatibility check against org.scala-lang:scala-library:2.12.0! Found 4 potential problems (filtered 655)
[error]  * static method empty()scala.collection.GenTraversable in interface scala.collection.immutable.Seq has a different result type in current version, where it is scala.collection.immutable.Seq rather than scala.collection.GenTraversable
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("scala.collection.immutable.Seq.empty")
[error]  * static method empty()scala.collection.immutable.Seq in interface scala.collection.immutable.Seq has a different result type in other version, where it is scala.collection.GenTraversable rather than scala.collection.immutable.Seq
[error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("scala.collection.immutable.Seq.empty")
[error]  * method empty()scala.collection.immutable.Seq in object scala.collection.immutable.Seq does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Seq.empty")
[error]  * method apply(scala.collection.Seq)scala.collection.immutable.Seq in object scala.collection.immutable.Seq does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.Seq.apply")
```

retronym replied:
Yes, in 2.12.x we would need to workaround this with:

```diff
--- a/src/library/scala/collection/generic/GenericCompanion.scala
+++ b/src/library/scala/collection/generic/GenericCompanion.scala
@@ -39,7 +39,10 @@ abstract class GenericCompanion[+CC[X] <: GenTraversable[X]] {
   /** An empty collection of type `$Coll[A]`
    *  @tparam A      the type of the ${coll}'s elements
    */
-  def empty[A]: CC[A] = newBuilder[A].result()
+  def empty[A]: CC[A] = {
+    if (this == immutable.Seq) Nil
+    else newBuilder[A].result()
+  }
```

retronym reviewed (changes requested):
Move the optimization into the inherited methods for binary compatibility.

mkeskells commented:
> Move the optimization into the inherited methods for binary compatibility.

Reworked as suggested. Also reworked `collection.Seq` and added a test for that as in 2.12 `Seq` == `collection.Seq`, so user code with `Seq.empty` refers to `collection.Seq.empty`, not `collection.immutable.seq.empty`
